src/components/activity-client-interface.js

```
@@ -0,0 +1,127 @@
+   * @param {!TrustedResourceUrl} url
+   * @param {?Object=} opt_args
+   */
+  constructor(iframe, url, opt_args) {
+    this.activityIframePort_ = new ActivityIframePort(iframe, url, opt_args);
+    /** @private @const {!Object<string, function(!{SerializedBuffer})>} */
+    this.callbackMap_ = {};
+  }
+
+  /**
+   *
+   * @param {string} pub_id
+   * @param {?AnalyticsContext} context
+   */
+  start(pub_id, context) {
+    this.message(
+        {'pubId': pub_id, 'context': context.toArray()});
+  }
+
+  /**
+   * @return {!ActivityMode}
+   */
+  getMode() {
+    return this.activityIframePort_.getMode();
+  }
+
+  /**
+   * Waits until the activity port is connected to the host.
+   * @return {!Promise}
+   */
+  connect() {
+    return this.activityIframePort_.connect();
+  }
+
+  disconnect() {
+    return this.activityIframePort_.disconnect();
+  }
+  /**
+   * @return {!Promise<!ActivityResult>}
+   */
+  acceptResult() {
+    return this.activityIframePort_.acceptResult();
+  }
+
+  /**
+   * @param {!Object} payload
+   */
+  message(payload) {
+    this.activityIframePort_.message(payload);
+  }
+
+  /**
+   * Registers a callback to receive messages from the host.
+   * @param {function(!Object)} callback
+   */
+  onMessage(callback) {
+    this.activityIframePort_.onMessage(callback);
+  }
+
+  messageChannel(opt_name) {
+    return this.activityIframePort_.messageChannel(opt_name);
+  }
+
+  whenReady() {
+    return this.activityIframePort_.whenReady();
+  }
+
+  onResizeRequest(callback) {
+    this.activityIframePort_.onResizeRequest(callback);
+  }
+
+  resized() {
+    this.activityIframePort_.resized();
+  }
+
+  /**
+   *
+   * @param {!SerializedBuffer} request
+   */
+  execute(request) {
+    this.message({req: request});
+  }
+
+  /**
+   * @param {string} responseType
+   * @param {function(!{!SerializedBuffer})} callback
+   */
+  on(responseType, callback) {
+    if (!this.callbackMap_[responseType]) {
```

If we use a single handler system (which is great), we should throw an error for duplicate handlers.